src/storage/mvcc/txn.rs

```diff
@@ -12,18 +12,99 @@
 // limitations under the License.
 
 use std::fmt;
-use storage::{Key, Value, Mutation};
+use std::collections::BTreeMap;
+use storage::Key;
 use storage::engine::{Engine, Snapshot, Modify, Cursor};
-use kvproto::mvccpb::{MetaLock, MetaLockType, MetaItem};
+use kvproto::kvpb::{Row, RowValue, Op, Mutation};
+use kvproto::mvccpb::{MetaLock, MetaItem, MetaColumn};
 use kvproto::kvrpcpb::Context;
 use super::meta::{Meta, FIRST_META_INDEX};
 use super::{Error, Result};
 
-fn meta_lock_type(mutation: &Mutation) -> MetaLockType {
-    match *mutation {
-        Mutation::Put(_) |
-        Mutation::Delete(_) => MetaLockType::ReadWrite,
-        Mutation::Lock(_) => MetaLockType::ReadOnly,
+trait MvccReader {
+    fn read(&mut self, key: &Key) -> Result<Option<Vec<u8>>>;
+
+    fn load_meta(&mut self, row: &Key, index: u64) -> Result<Meta> {
+        let meta = match try!(self.read(&row.append_ts(index))) {
+            Some(x) => try!(Meta::parse(&x)),
+            None => Meta::new(),
+        };
+        Ok(meta)
+    }
+
+    fn get_row(&mut self, row: Row, ts: u64) -> Result<RowValue> {
+        let row_key = Key::from_raw(row.get_row_key());
+        let meta = try!(self.load_meta(&row_key, FIRST_META_INDEX));
+        self.get_row_with_first_meta(&row_key, row, &meta, ts)
+    }
+
+    fn get_row_with_first_meta(&mut self,
+                               row_key: &Key,
+                               mut row: Row,
+                               first_meta: &Meta,
+                               ts: u64)
+                               -> Result<RowValue> {
+        // Check for locks that signal concurrent writes.
+        if let Some(lock) = first_meta.get_lock() {
+            if lock.get_start_ts() <= ts {
+                // There is a pending lock. Client should wait or clean it.
+                return Err(Error::KeyIsLocked {
+                    key: row.get_row_key().to_vec(),
+                    primary: lock.get_primary_key().to_vec(),
+                    ts: lock.get_start_ts(),
+                });
+            }
+        }
+        let mut row_value = RowValue::new();
+        row_value.set_row_key(row.take_row_key());
+        let mut pending_cols = BTreeMap::new();
+        for col in row.get_columns() {
+            pending_cols.insert(col.to_vec(), true);
+        }
+        // Find the latest write below our start timestamp.
+        if let Some(x) = first_meta.iter_items().find(|x| x.get_commit_ts() <= ts) {
```

Comment:
find() returns the first element satisfying the specified predicate, some cols may get wrong version data or get no data